posts/2023-05-19-hibernate-reactive-gotchas.adoc

```diff
@@ -0,0 +1,215 @@
+
+== How to prevent this?
+
+=== If your own code evaluates Uni's in parallel
+
+If your own code evaluates Uni's in parallel, you have several options:
+
+The most obvious option is to use `.usingConcurrencyOf(1)` when using `Uni.join()` or `Uni.combine()`, for example:
+
+   Uni.join().all(one, two).usingConcurrencyOf(1); // Will evaluate `one` and `two` sequentially.
+
+Another way is two use `flatMap` for more explicit sequential evaluation:
+
+    one.flatMap(x -> two.map(y -> x + " " + y)); // Will evaluate `one` and `two` sequentially.
+
+=== If someone else's code evaluates Uni's in parallel
+
+Sometimes you pass ``Uni``s from Hibernate Reactive to some library for further processing. In that case, you may not have control over how this library evaluates the ``Uni``s.
+
+This bit us in the past, when using Hibernate Reactive together with Smallrye GraphQL: https://github.com/quarkusio/quarkus/issues/32870
+
+In this scenario, the GraphQL engine decides to evaluate Uni's in parallel (which it _should_, according to the GraphQL spec!), which is not allowed for Hibernate ``Uni``'s.
+
+In the next sections we'll dive a little deeper into the fundamental problem here, solutions of other libraries, and what we can do in our situation.
+
+== What's the problem here?
+
+In my opinion, the fundamental problem is that Hibernate Reactive doesn't return 'regular' ``Uni``'s, but ``Uni``'s with an additional instruction manual. The instruction manual (_don't evaluate them in parallel!_) doesn't show up in the type, so developers won't know about them until they read the manual. Worse, other libraries won't know about them either and violate their rules.
+
+=== Another approach: Doobie
+
+https://tpolecat.github.io/doobie/[Doobie] is a popular database access library for Scala, which has solved this by using a specific type `ConnectionIO`, which _does not allow parallel composition_, but only sequential composition. You _can_ convert from this type to a more general type (somewhat like `Uni`), but at that point you define the transaction boundary.
+
+So the entire instruction manual of _you can not run database work on the same transaction in parallel_, which is similar to the constraint of Hibernate, is encoded in the type system. Beautiful.
+
+== Making your Uni's safe
+
+So, the problem is that Hibernate ``Uni``'s should not be evaluated in parallel, and that libraries that consume the ``Uni``'s are not aware of this restriction.
+
+So naturally, we could ask: can we remove this constraint, and make our ``Uni``s safe for parallel execution?
+
+The answer is that we can, and it's not terribly complicated.
+
+What we want to do is to shift the responsibility for sequential execution from outside the ``Uni``'s to _inside_ the ``Uni``'s. So that if two Hibernate Reactive ``Uni``s are run in parallel, one of them will wait with executing the actual Hibernate work until the other is done.
```

Comment:
```suggestion
What we want to do is to shift the responsibility for sequential execution from outside the ``Uni``s to _inside_ the ``Uni``s. So that if two Hibernate Reactive ``Uni``s are run in parallel, one of them will wait with executing the actual Hibernate work until the other is done.
```